collector/zfs_linux.go

```diff
@@ -27,31 +27,110 @@ import (
 )
 
 const (
-	zfsArcstatsProcpath = "spl/kstat/zfs/arcstats"
+	zfsProcpathBase      = "spl/kstat/zfs/"
+	zfsArcstatsExt       = "arcstats"
+	zfsDmuTxExt          = "dmu_tx"
+	zfsFmExt             = "fm"
+	zfsFetchstatsExt     = "zfetchstats"
+	zfsVdevCacheStatsExt = "vdev_cache_stats"
+	zfsXuioStatsExt      = "xuio_stats"
+	zfsZilExt            = "zil"
 )
 
-func (c *zfsCollector) openArcstatsFile() (file *os.File, err error) {
-	file, err = os.Open(procFilePath(zfsArcstatsProcpath))
+func (c *zfsCollector) openProcFile(path string) (file *os.File, err error) {
+	file, err = os.Open(procFilePath(path))
 	if err != nil {
-		log.Debugf("Cannot open %q for reading. Is the kernel module loaded?", procFilePath(zfsArcstatsProcpath))
+		log.Debugf("Cannot open %q for reading. Is the kernel module loaded?", procFilePath(path))
 		err = zfsNotAvailableError
 	}
 	return
 }
 
 func (c *zfsCollector) updateArcstats(ch chan<- prometheus.Metric) (err error) {
-	file, err := c.openArcstatsFile()
+	file, err := c.openProcFile(filepath.Join(zfsProcpathBase, zfsArcstatsExt))
 	if err != nil {
 		return err
 	}
 	defer file.Close()
 
-	return c.parseArcstatsProcfsFile(file, func(s zfsSysctl, v zfsMetricValue) {
+	return c.parseProcfsFile(file, zfsArcstatsExt, func(s zfsSysctl, v zfsMetricValue) {
 		ch <- c.constSysctlMetric(arc, s, v)
 	})
 }
 
-func (c *zfsCollector) parseArcstatsProcfsFile(reader io.Reader, handler func(zfsSysctl, zfsMetricValue)) (err error) {
+func (c *zfsCollector) updateZfetchstats(ch chan<- prometheus.Metric) (err error) {
+	file, err := c.openProcFile(filepath.Join(zfsProcpathBase, zfsFetchstatsExt))
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	return c.parseProcfsFile(file, zfsFetchstatsExt, func(s zfsSysctl, v zfsMetricValue) {
+		ch <- c.constSysctlMetric(zfetch, s, v)
+	})
+}
+
+func (c *zfsCollector) updateZil(ch chan<- prometheus.Metric) (err error) {
```

Comment:
Those function all do pretty much the same. Some DRY would be nice. Can't we iterate over a map with those files for example, instead of having a function for each?